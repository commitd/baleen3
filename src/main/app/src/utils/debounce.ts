/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
/* eslint-disable @typescript-eslint/no-explicit-any */
export type Method = (...args: any[]) => void

export function debounce<F extends Method>(
  func: F,
  wait = 200
): (this: ThisParameterType<F>, ...args: Parameters<F>) => void {
  let timeoutId: ReturnType<typeof setTimeout> | undefined

  return function (this: ThisParameterType<F>, ...args: Parameters<F>): void {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const context = this

    if (timeoutId !== undefined) {
      clearTimeout(timeoutId)
    }
    timeoutId = setTimeout(() => {
      timeoutId = undefined
      func.apply(context, args)
    }, wait)
  }
}
