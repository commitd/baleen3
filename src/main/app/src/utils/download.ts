/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const downloadJson = (name: string, data: any): void => {
  download(
    name + '.json',
    'data:text/json;charset=utf-8,',
    JSON.stringify(data, null, 2)
  )
}

export const download = (
  name: string,
  mediaType: string,
  data: string
): void => {
  const dataStr = mediaType + encodeURIComponent(data)
  const downloadAnchorNode = document.createElement('a')
  downloadAnchorNode.setAttribute('href', dataStr)
  downloadAnchorNode.setAttribute('download', name)
  document.body.appendChild(downloadAnchorNode) // required for firefox
  downloadAnchorNode.click()
  downloadAnchorNode.remove()
}
