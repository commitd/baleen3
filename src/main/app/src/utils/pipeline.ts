/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
import Ajv from 'ajv'
import { v4 } from 'uuid'
import {
  Annot8ComponentDescriptor,
  ComponentInfo,
  ComponentMap,
  ComponentMetadata,
  PipelineDescriptor,
  PipelineEditDescriptor,
  PipelineViewDescriptor,
  ServerDetails,
  SettingsMap,
  SettingsSchema,
} from '../types'

const ajv = new Ajv({ useDefaults: true })
// eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-return
const uuid = (): string => v4()

export const isSettingsValid = (
  schema?: SettingsSchema,
  settings?: object
): boolean => {
  if (schema === undefined) {
    // No settings schema, assume valid
    return true
  }

  try {
    const jsonSchema = JSON.parse(schema.jsonSchema) as object
    const validator = ajv.compile(jsonSchema)
    const valid = validator(settings) as boolean
    if (!valid) {
      let message = 'JSON not valid:'
      if (!(validator.errors == null)) {
        validator.errors.forEach((error: Ajv.ErrorObject): void => {
          message += error.message
        })
      }
      console.debug(message)
    }
    return valid
  } catch (e) {
    console.debug(e)
    return false
  }
}

export const defaultSettings = (
  schema?: SettingsSchema
): object | undefined => {
  if (schema === undefined) {
    // No settings schema, assume no settings
    return undefined
  }

  try {
    const jsonSchema = JSON.parse(schema.jsonSchema) as object
    const validator = ajv.compile(jsonSchema)
    const defaultData = {}
    const valid = validator({}) as boolean
    if (valid) {
      return defaultData
    }
  } catch (e) {
    console.debug(e)
  }
  return undefined
}

export const isAllComponentsValid = (
  components: ComponentMetadata[]
): boolean => {
  return components.find((c) => !c.valid) === undefined
}

export const isPipelineValid = (
  nameValid: boolean,
  pipeline: Readonly<PipelineViewDescriptor>
): string[] => {
  const errors = []
  if (!nameValid) {
    errors.push('The pipeline name must be unique')
  }
  if (pipeline.sourceOrder.length === 0) {
    errors.push('A pipeline must have at least one source')
  }
  if (pipeline.processorOrder.length === 0) {
    errors.push('A pipeline must have at least one processor')
  }
  if (!isAllComponentsValid(Object.values(pipeline.components))) {
    errors.push('All components must be valid')
  }

  return errors
}

export const createComponent =
  (settings: SettingsMap, types: ComponentMap, configured = true) =>
  (descriptor: Annot8ComponentDescriptor): ComponentMetadata => {
    const id = uuid()
    const type = Object.keys(descriptor)[0]
    const info: ComponentInfo | undefined = types[`${type}`]
    const schema: SettingsSchema | undefined =
      info !== undefined && info.settingsClass !== undefined
        ? settings[info.settingsClass]
        : undefined

    if (schema !== undefined && descriptor[`${type}`].settings === undefined) {
      descriptor[`${type}`].settings = defaultSettings(schema)
    }

    return {
      id,
      descriptor,
      info,
      schema,
      valid:
        info !== undefined &&
        isSettingsValid(schema, descriptor[`${type}`].settings),
      configured,
    }
  }

export const createPipelineViewDescriptor = (
  pipeline: PipelineDescriptor,
  serverDetails: ServerDetails
): PipelineViewDescriptor => {
  const components: { [key: string]: ComponentMetadata } = {}
  const sourceOrder: string[] = []
  const processorOrder: string[] = []

  pipeline.sources
    .map(createComponent(serverDetails.settings, serverDetails.sources))
    .forEach((component) => {
      components[component.id] = component
      sourceOrder.push(component.id)
    })

  pipeline.processors
    .map(createComponent(serverDetails.settings, serverDetails.processors))
    .forEach((component) => {
      components[component.id] = component
      processorOrder.push(component.id)
    })

  return {
    name: pipeline.name,
    description: pipeline.description,
    sourceOrder,
    processorOrder,
    components,
    errorConfiguration: pipeline.errorConfiguration,
  }
}

export const createPipelineEditDescriptor = (
  pipeline: PipelineDescriptor,
  orderer: string,
  usedNames: string[],
  serverDetails: ServerDetails
): PipelineEditDescriptor => {
  const pipelineView = createPipelineViewDescriptor(pipeline, serverDetails)

  const nameValid = !usedNames.includes(pipeline.name)

  return {
    ...pipelineView,
    orderer,
    nameValid,
    errors: isPipelineValid(nameValid, pipelineView),
  }
}
