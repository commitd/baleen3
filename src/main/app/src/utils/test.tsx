/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
import { StylesProvider } from '@material-ui/styles'
import {
  createHistory,
  createMemorySource,
  History,
  LocationProvider,
} from '@reach/router'
import { render, RenderOptions, RenderResult } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { GenerateId } from 'jss'
import React from 'react'
import { ThemeProvider } from '../components/ThemeProvider/ThemeProvider'

const generateClassName: GenerateId = (rule, styleSheet) => {
  const prefix =
    styleSheet === undefined || styleSheet.options.classNamePrefix === undefined
      ? ''
      : styleSheet.options.classNamePrefix
  return `${prefix}-${rule.key}`
}

const LightTheme: React.FC<{ readonly children?: React.ReactNode }> = ({
  children,
}) => (
  <StylesProvider generateClassName={generateClassName}>
    <ThemeProvider choice="light">{children}</ThemeProvider>
  </StylesProvider>
)

const DarkTheme: React.FC<{ readonly children?: React.ReactNode }> = ({
  children,
}) => (
  <StylesProvider generateClassName={generateClassName}>
    <ThemeProvider choice="dark">{children}</ThemeProvider>
  </StylesProvider>
)

export const renderPlain = render

export const renderLight = (
  ui: Readonly<React.ReactElement>,
  options?: Omit<RenderOptions, 'queries'>
): RenderResult => render(ui, { wrapper: LightTheme, ...options })

export const renderDark = (
  ui: Readonly<React.ReactElement>,
  options?: Omit<RenderOptions, 'queries'>
): RenderResult => render(ui, { wrapper: DarkTheme, ...options })

export function renderWithRouter(
  ui: Readonly<React.ReactElement>,
  route = '/',
  options?: Omit<RenderOptions, 'queries'>,
  delegate = renderLight
): RenderResult & { history: History } {
  const history = createHistory(createMemorySource(route))
  return {
    ...delegate(
      <LocationProvider history={history}>{ui}</LocationProvider>,
      options
    ),
    history,
  }
}
// re-export everything
export * from '@testing-library/react'
// override render method
export { renderLight as render }
export { userEvent }
