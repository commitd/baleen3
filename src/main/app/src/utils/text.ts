/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
export const javaClassToJavaName = (javaName: string): string => {
  const split = javaName.split('.')
  const lastValue = split.pop()
  return lastValue === undefined ? '' : lastValue
}

export const javaNameToReadable = (javaName: string): string => {
  const separated = javaName
    .replace(/([A-Z][A-Z])/g, ' $1')
    .replace(/([A-Z][a-z$])/g, ' $1')
    .replace(/[$]/g, '')
  return separated.trim()
}

export const javaToReadable = (input = ''): string =>
  javaNameToReadable(javaClassToJavaName(input))

export const artifactToReadable = (artifact: string): string => {
  const reduced = artifact.replace('annot8-components-', '')
  const split = reduced.split('-')
  return split.map((value) => value[0].toUpperCase() + value.slice(1)).join(' ')
}

export const parseSearch = (search: string): string[] =>
  search
    .split(/[,;]\s*/)
    .map(standardize)
    .filter((term) => term !== '')

export const standardize = (s: string): string =>
  s.trim().toLowerCase().replace(/\s+/g, ' ')
