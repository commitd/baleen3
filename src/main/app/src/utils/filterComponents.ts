/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
import { ComponentInfo, ComponentMap } from '../types'
import { parseSearch, standardize } from './text'

export const filterComponents = (
  map: ComponentMap,
  search: string
): Array<ComponentInfo> => {
  const values = Object.values(map)
  if (search.trim() === '') {
    return values
  }

  const searchTerms = parseSearch(search)

  return values.filter(
    (value) =>
      filterBy('artifact', value, searchTerms) ||
      filterBy('description', value, searchTerms) ||
      filterByName(value, searchTerms) ||
      filterByTags(value, searchTerms)
  )
}

function filterBy(
  key: keyof Omit<ComponentInfo, 'tags'>,
  value: ComponentInfo,
  terms: string[]
): boolean {
  // false positive security/detect-object-injection
  // eslint-disable-next-line
  const property = value[key]
  if (property === undefined) {
    return false
  }
  return terms.some((term) => standardize(property).includes(term))
}

function filterByName(value: ComponentInfo, terms: string[]): boolean {
  const property = standardize(value.name)
  return terms.some((term) => property.includes(term))
}

function filterByTags(value: ComponentInfo, terms: string[]): boolean {
  const tags = value.tags?.map((tag) => tag.toLowerCase())
  return terms.some((term) => tags?.includes(term))
}
