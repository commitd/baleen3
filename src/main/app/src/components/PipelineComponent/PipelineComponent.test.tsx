/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
import React from 'react'
import { PipelineComponent } from '.'
import {
  exampleProcessorMetadata,
  exampleSourceMetadata,
} from '../../types/examples'
import { act, render, renderDark, userEvent } from '../../utils/test'

it('renders without error', () => {
  const { asFragment } = render(
    <PipelineComponent type="Processor" descriptor={exampleSourceMetadata} />
  )
  expect(asFragment()).toMatchSnapshot()
})

it('renders dark without error', () => {
  const { asFragment } = renderDark(
    <PipelineComponent type="Processor" descriptor={exampleSourceMetadata} />
  )
  expect(asFragment()).toMatchSnapshot()
})

it('renders without settings', () => {
  const descriptor = Object.assign({}, exampleProcessorMetadata, {
    schema: undefined,
  })

  const { asFragment } = render(
    <PipelineComponent type="Processor" descriptor={descriptor} />
  )
  expect(asFragment()).toMatchSnapshot()
})

it('renders without componentInfo', () => {
  const descriptor = Object.assign({}, exampleProcessorMetadata, {
    info: undefined,
  })

  const { asFragment } = render(
    <PipelineComponent type="Processor" descriptor={descriptor} />
  )
  expect(asFragment()).toMatchSnapshot()
})

it('renders when editable componentInfo', () => {
  const setName = jest.fn()
  const setSettings = jest.fn()
  const onDelete = jest.fn()
  const { getByRole } = render(
    <PipelineComponent
      type="Processor"
      descriptor={exampleProcessorMetadata}
      setName={setName}
      setSettings={setSettings}
      onDelete={onDelete}
    />
  )
  act(() => {
    userEvent.click(getByRole('button', { name: /delete/i }))
  })
  expect(onDelete).toHaveBeenCalled()
})

it('Shows error when JSON does not parse', () => {
  const errorSpy = jest.spyOn(console, 'error').mockImplementation()
  const logSpy = jest.spyOn(console, 'log').mockImplementation()
  const setName = jest.fn()
  const setSettings = jest.fn()
  const onDelete = jest.fn()

  const descriptor = Object.assign({}, exampleProcessorMetadata, {
    schema: {
      jsonSchema: '{"$schema":"This does not parse}',
    },
  })

  try {
    const { getByText } = render(
      <PipelineComponent
        type="Processor"
        descriptor={descriptor}
        setName={setName}
        setSettings={setSettings}
        onDelete={onDelete}
      />
    )

    getByText('Error parsing Settings')
    expect(errorSpy).toHaveBeenCalled()
    expect(logSpy).toHaveBeenCalled()
  } finally {
    errorSpy.mockRestore()
    logSpy.mockRestore()
  }
})
