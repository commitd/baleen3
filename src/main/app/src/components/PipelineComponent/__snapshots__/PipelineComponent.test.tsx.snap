// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders dark without error 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
    >
      <div
        tabindex="-1"
      >
        <div
          class="MuiPaper-root MuiCard-root WithStyles(ForwardRef(Card))-root WithStyles(ForwardRef(Card))-root MuiPaper-elevation1 MuiPaper-rounded"
          tabindex="0"
        >
          <div
            class="MuiCardHeader-root WithStyles(ForwardRef(CardHeader))-root WithStyles(ForwardRef(CardHeader))-root"
          >
            <div
              class="MuiCardHeader-content"
            >
              <span
                class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
              >
                Files
              </span>
              <span
                class="MuiTypography-root MuiCardHeader-subheader MuiTypography-caption MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                File System Source
              </span>
            </div>
            <div
              class="MuiCardHeader-action"
            >
              <div
                style="opacity: 0; visibility: hidden;"
              >
                <button
                  aria-label="view settings"
                  class="MuiButtonBase-root MuiIconButton-root WithNoOpener-root WithNoOpener-root"
                  tabindex="0"
                  title="View Settings"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19.44 12.99l-.01.02c.04-.33.08-.67.08-1.01 0-.34-.03-.66-.07-.99l.01.02 2.44-1.92-2.43-4.22-2.87 1.16.01.01c-.52-.4-1.09-.74-1.71-1h.01L14.44 2H9.57l-.44 3.07h.01c-.62.26-1.19.6-1.71 1l.01-.01-2.88-1.17-2.44 4.22 2.44 1.92.01-.02c-.04.33-.07.65-.07.99 0 .34.03.68.08 1.01l-.01-.02-2.1 1.65-.33.26 2.43 4.2 2.88-1.15-.02-.04c.53.41 1.1.75 1.73 1.01h-.03L9.58 22h4.85s.03-.18.06-.42l.38-2.65h-.01c.62-.26 1.2-.6 1.73-1.01l-.02.04 2.88 1.15 2.43-4.2s-.14-.12-.33-.26l-2.11-1.66zM12 15.5c-1.93 0-3.5-1.57-3.5-3.5s1.57-3.5 3.5-3.5 3.5 1.57 3.5 3.5-1.57 3.5-3.5 3.5z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
          <div
            class="MuiCardContent-root"
          >
            <div
              class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
            >
              <div
                class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
              >
                <p
                  class="MuiTypography-root WithStyles(ForwardRef(Typography))-root WithStyles(ForwardRef(Typography))-root WrappedTypography-root WrappedTypography-root MuiTypography-body1"
                >
                  Provides items from the local file system
                </p>
              </div>
              <div
                class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders without componentInfo 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
    >
      <div
        tabindex="-1"
      >
        <div
          class="MuiPaper-root MuiCard-root WithStyles(ForwardRef(Card))-root WithStyles(ForwardRef(Card))-root MuiPaper-elevation1 MuiPaper-rounded"
          tabindex="0"
        >
          <div
            class="MuiPaper-root MuiAlert-root MuiAlert-filledError MuiPaper-elevation0"
            role="alert"
          >
            <div
              class="MuiAlert-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
            </div>
            <div
              class="MuiAlert-message"
            >
              Component type not available
            </div>
          </div>
          <div
            class="MuiCardHeader-root WithStyles(ForwardRef(CardHeader))-root WithStyles(ForwardRef(CardHeader))-root"
          >
            <div
              class="MuiCardHeader-content"
            >
              <span
                class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
              >
                Geo
              </span>
              <span
                class="MuiTypography-root MuiCardHeader-subheader MuiTypography-caption MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                Missing
              </span>
            </div>
            <div
              class="MuiCardHeader-action"
            >
              <div
                style="opacity: 0; visibility: hidden;"
              >
                <button
                  aria-label="view settings"
                  class="MuiButtonBase-root MuiIconButton-root WithNoOpener-root WithNoOpener-root"
                  tabindex="0"
                  title="View Settings"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19.44 12.99l-.01.02c.04-.33.08-.67.08-1.01 0-.34-.03-.66-.07-.99l.01.02 2.44-1.92-2.43-4.22-2.87 1.16.01.01c-.52-.4-1.09-.74-1.71-1h.01L14.44 2H9.57l-.44 3.07h.01c-.62.26-1.19.6-1.71 1l.01-.01-2.88-1.17-2.44 4.22 2.44 1.92.01-.02c-.04.33-.07.65-.07.99 0 .34.03.68.08 1.01l-.01-.02-2.1 1.65-.33.26 2.43 4.2 2.88-1.15-.02-.04c.53.41 1.1.75 1.73 1.01h-.03L9.58 22h4.85s.03-.18.06-.42l.38-2.65h-.01c.62-.26 1.2-.6 1.73-1.01l-.02.04 2.88 1.15 2.43-4.2s-.14-.12-.33-.26l-2.11-1.66zM12 15.5c-1.93 0-3.5-1.57-3.5-3.5s1.57-3.5 3.5-3.5 3.5 1.57 3.5 3.5-1.57 3.5-3.5 3.5z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
          <div
            class="MuiCardContent-root"
          >
            <div
              class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
            >
              <div
                class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
              >
                <p
                  class="MuiTypography-root WithStyles(ForwardRef(Typography))-root WithStyles(ForwardRef(Typography))-root WrappedTypography-root WrappedTypography-root MuiTypography-body1"
                />
              </div>
              <div
                class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders without error 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
    >
      <div
        tabindex="-1"
      >
        <div
          class="MuiPaper-root MuiCard-root WithStyles(ForwardRef(Card))-root WithStyles(ForwardRef(Card))-root MuiPaper-elevation1 MuiPaper-rounded"
          tabindex="0"
        >
          <div
            class="MuiCardHeader-root WithStyles(ForwardRef(CardHeader))-root WithStyles(ForwardRef(CardHeader))-root"
          >
            <div
              class="MuiCardHeader-content"
            >
              <span
                class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
              >
                Files
              </span>
              <span
                class="MuiTypography-root MuiCardHeader-subheader MuiTypography-caption MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                File System Source
              </span>
            </div>
            <div
              class="MuiCardHeader-action"
            >
              <div
                style="opacity: 0; visibility: hidden;"
              >
                <button
                  aria-label="view settings"
                  class="MuiButtonBase-root MuiIconButton-root WithNoOpener-root WithNoOpener-root"
                  tabindex="0"
                  title="View Settings"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19.44 12.99l-.01.02c.04-.33.08-.67.08-1.01 0-.34-.03-.66-.07-.99l.01.02 2.44-1.92-2.43-4.22-2.87 1.16.01.01c-.52-.4-1.09-.74-1.71-1h.01L14.44 2H9.57l-.44 3.07h.01c-.62.26-1.19.6-1.71 1l.01-.01-2.88-1.17-2.44 4.22 2.44 1.92.01-.02c-.04.33-.07.65-.07.99 0 .34.03.68.08 1.01l-.01-.02-2.1 1.65-.33.26 2.43 4.2 2.88-1.15-.02-.04c.53.41 1.1.75 1.73 1.01h-.03L9.58 22h4.85s.03-.18.06-.42l.38-2.65h-.01c.62-.26 1.2-.6 1.73-1.01l-.02.04 2.88 1.15 2.43-4.2s-.14-.12-.33-.26l-2.11-1.66zM12 15.5c-1.93 0-3.5-1.57-3.5-3.5s1.57-3.5 3.5-3.5 3.5 1.57 3.5 3.5-1.57 3.5-3.5 3.5z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
          <div
            class="MuiCardContent-root"
          >
            <div
              class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
            >
              <div
                class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
              >
                <p
                  class="MuiTypography-root WithStyles(ForwardRef(Typography))-root WithStyles(ForwardRef(Typography))-root WrappedTypography-root WrappedTypography-root MuiTypography-body1"
                >
                  Provides items from the local file system
                </p>
              </div>
              <div
                class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders without settings 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
    >
      <div
        tabindex="-1"
      >
        <div
          class="MuiPaper-root MuiCard-root WithStyles(ForwardRef(Card))-root WithStyles(ForwardRef(Card))-root MuiPaper-elevation1 MuiPaper-rounded"
          tabindex="0"
        >
          <div
            class="MuiCardHeader-root WithStyles(ForwardRef(CardHeader))-root WithStyles(ForwardRef(CardHeader))-root"
          >
            <div
              class="MuiCardHeader-content"
            >
              <span
                class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
              >
                Geo
              </span>
              <span
                class="MuiTypography-root MuiCardHeader-subheader MuiTypography-caption MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                MGRS
              </span>
            </div>
            <div
              class="MuiCardHeader-action"
            >
              <div
                style="opacity: 0; visibility: hidden;"
              >
                <button
                  aria-label="view settings"
                  class="MuiButtonBase-root MuiIconButton-root WithNoOpener-root WithNoOpener-root MuiIconButton-disabled MuiButtonBase-disabled"
                  disabled=""
                  tabindex="-1"
                  title="View Settings"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19.44 12.99l-.01.02c.04-.33.08-.67.08-1.01 0-.34-.03-.66-.07-.99l.01.02 2.44-1.92-2.43-4.22-2.87 1.16.01.01c-.52-.4-1.09-.74-1.71-1h.01L14.44 2H9.57l-.44 3.07h.01c-.62.26-1.19.6-1.71 1l.01-.01-2.88-1.17-2.44 4.22 2.44 1.92.01-.02c-.04.33-.07.65-.07.99 0 .34.03.68.08 1.01l-.01-.02-2.1 1.65-.33.26 2.43 4.2 2.88-1.15-.02-.04c.53.41 1.1.75 1.73 1.01h-.03L9.58 22h4.85s.03-.18.06-.42l.38-2.65h-.01c.62-.26 1.2-.6 1.73-1.01l-.02.04 2.88 1.15 2.43-4.2s-.14-.12-.33-.26l-2.11-1.66zM12 15.5c-1.93 0-3.5-1.57-3.5-3.5s1.57-3.5 3.5-3.5 3.5 1.57 3.5 3.5-1.57 3.5-3.5 3.5z"
                      />
                    </svg>
                  </span>
                </button>
              </div>
            </div>
          </div>
          <div
            class="MuiCardContent-root"
          >
            <div
              class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
            >
              <div
                class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
              >
                <p
                  class="MuiTypography-root WithStyles(ForwardRef(Typography))-root WithStyles(ForwardRef(Typography))-root WrappedTypography-root WrappedTypography-root MuiTypography-body1"
                >
                  Extract MGRS coordinates, optionally ignoring MGRS coordinates that could be dates
                </p>
              </div>
              <div
                class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
